### ISSUE TITLE
Allow configurable time for ES output

### ISSUE THREAD

**Issue by markwalkom** [elastic/beats#921]
[Here](https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-configuration-details.html#elasticsearch-output) we state;
```
# Optional index name. The default is "filebeat" and generates
    # [filebeat-]YYYY.MM.DD keys.
    index: "filebeat"
```

But we don't allow the `YYYY.MM.DD` to be configurable.

This isn't ideal as users cannot do anything but daily indices without putting Logstash in the pipeline.

**Comment by ruflin** [comment]
An older discussion about this can be found here: https://github.com/elastic/libbeat/issues/318 It's a good time to close the older one and reference to this one here.

Somehow related: https://github.com/elastic/beats/issues/574

**Comment by ruflin** [comment]
@markwalkom Is the goal to just make it possible to not have a pattern, or that the output would recognize different output patterns? Can you give some examples?

**Comment by markwalkom** [comment]
Sure.

Hourly, weekly, monthly indices. Not non-time based though.

**Comment by ruflin** [comment]
That would mean it has the following logic (abstract code)

```
var index = config.indexname
var pattern = config.pattern

// Pattern not set falls back to default pattern
if pattern == nil {
	pattern = "YYYY.MM.DD"
}


// Empty pattern means pattern is not used
if pattern != "" {
	
	// Append pattern to index based on timestamp
	pattern = ParsePattern(config.pattern, time.Now())
	index = index + "-" + $PATTERN
}
```
Would that work, assuming we can interpret common pattern with `ParsePattern`?

**Comment by markwalkom** [comment]
Looks OK.

You can do this with LS, not sure how go handles it!

**Comment by gr0** [comment]
Would it be possible to have the index and pattern concatenation part also configurable, so that we could have indices like index_pattern, not only index-pattern?

**Comment by markwalkom** [comment]
@gr0 you mean `_` and `-` as the separator?

**Comment by gr0** [comment]
@markwalkom yes, so one could actually have the index pattern that matches the rest of the sources and not ending up with multiple index name patterns or having to modify the rest of the setup to include beats. 

For example, if would be nice, to be able to have patterns like [filebeat_]YYYY-MM-DD or [filebeat_]YYYY.MM.DD depending on the needs. Of course with some nice default if the user doesn't want to configure the pattern and just want beats to take care of that.

What do you think?

**Comment by UserNotFound** [comment]
I'd vote for duplicating how logstash parses this option, at a minimum.

**Comment by ruflin** [comment]
As a reference here is the way LS does it: https://www.elastic.co/guide/en/logstash/current/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-index

@UserNotFound You mention "at a minimum". What are the limitations for you with the LS approach?

**Comment by UserNotFound** [comment]
@ruflin I have not encountered any limitations, I would just like to be able to name all of my indexes consistently across my environment.

**Comment by ruflin** [comment]
@UserNotFound Ok. I was asking because if we implement this similar to LS it would be good to know if there are some limitations with the LS approach and we could improve on top of it.

**Comment by anhlqn** [comment]
Just gave winlogbeat a try today since we are using Solarwinds LEM to gather Windows events. Looking forward for this enhancement to be implemented because:

- I want to limit the number of index/shards for windows event log indexes. Even with 2 shards per indexes, they are still a lot
- Not having to add another layer of LS would save some resources, redundancy issue, and headead ofc

**Comment by ruflin** [comment]
@markwalkom Could this one help here? https://github.com/elastic/elasticsearch/issues/17814

**Comment by markwalkom** [comment]
Yes, but only if the user uses the ingest node.

**Comment by ruflin** [comment]
@markwalkom So either ingest node or LS can be used, so question is if this feature would still be needed on the beats side (if we not talk about older ES versions).

**Comment by markwalkom** [comment]
But if you use direct to ES this still matters, so I think it's still a
good option to have.

**Comment by ruflin** [comment]
There is now also a new "rollover" API which is also interesting here: https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-rollover-index.html

**Comment by tsg** [comment]
Implemented in #2119.
